File: ssd_ros/scripts/sub_camera.py
#!/usr/bin/env python
#coding:utf-8
import rospy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

input_shape = (300, 300, 3)

def imageCallback(image_msg):
    
    # rospy.loginfo("Subscribe Image data")
    try:
        cv_image = CvBridge().imgmsg_to_cv2(image_msg,"bgr8")
    except CvBridgeError as e:
        print (e)
    
    (rows, cols, channels) = cv_image.shape
    cv_image_width  = cols
    cv_image_height = rows
    cv_image_dar    = cv_image_width / cv_image_height;

    print "Width:%d Height:%d" % (cv_image_width, cv_image_height)

    im_size = (input_shape[0], input_shape[1])
    resized = cv2.resize(cv_image, im_size)
    rgb = cv2.cvtColor(resized, cv2.COLOR_BGR2RGB)
    to_draw = cv2.resize(resized,(int(input_shape[0]*cv_image_dar),input_shape[1]))

    cv2.imshow("ssd_ros",cv_image)
    cv2.imshow("resized",to_draw)
    cv2.waitKey(10)

def main():
    rospy.init_node("sub_camera")

    sub_image = rospy.Subscriber("/usb_cam/image_raw",Image,imageCallback)
    try:
        rospy.spin()
    except KeyboardInterrupt:
        print ("Shutting down")
    cv2.destroyAllWindows()

if __name__ == '__main__':
    main()

File: ssd_ros/src/sub_camera.cpp
#include<ros/ros.h>
#include<sensor_msgs/Image.h>
#include<opencv2/opencv.hpp>
#include<cv_bridge/cv_bridge.h>

using namespace std;

//ros::Publisher pub_image;

sensor_msgs::Image image_;
void imageCallback(const sensor_msgs::ImageConstPtr& image_msg)
{
    image_ = *image_msg;
    //pub_image.publish(image_);
    
    cv_bridge::CvImageConstPtr cv_img_ptr;
    try{
        cv_img_ptr = cv_bridge::toCvShare(image_msg,"bgr8");
    }catch (cv_bridge::Exception& e){
        ROS_ERROR("cv_bridge exception: %s", e.what());
    }

    // cv::Mat image(cv_img_ptr->image.rows, cv_img_ptr->image.cols, cv_img_ptr->image.type());
    // image = cv_bridge::toCvShare(image_msg)->image;

    cv::Mat image = cv_img_ptr->image;
    cv::imshow("SSD result", image);
    cv::waitKey(10);
}

int main(int argc, char** argv)
{
    ros::init(argc,argv,"sub_camera");
    ros::NodeHandle n;

    printf("subscribe camera data\n");

    ros::Subscriber sub_image = n.subscribe("usb_cam/image_raw",10,imageCallback);

    //pub_image = n.advertise<sensor_msgs::Image>("ssd_ros/image_raw",10);

    ros::Rate rate(20);
    while(ros::ok())
    {
        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}
